src/simulators/ch/ch_state.hpp

```diff
@@ -0,0 +1,787 @@
+  // Three-qubit gates
+  {"ccx", Gates::ccx},    // Controlled-CX gate (Toffoli)
+  {"ccz", Gates::ccz}     // Constrolled-CZ gate (H3 Toff H3)
+});
+
+const stringmap_t<Snapshots> State::snapshotset_({
+  {"state", Snapshots::state},
+  {"statevector", Snapshots::statevector},
+  {"probabilities", Snapshots::probabilities},
+  {"memory", Snapshots::cmemory},
+  {"register", Snapshots::cregister}
+});
+
+//-------------------------------------------------------------------------
+// Implementation: Initialisation and Config
+//-------------------------------------------------------------------------
+
+void State::initialize_qreg(uint_t num_qubits)
+{
+  BaseState::qreg_.initialize(num_qubits);
+  BaseState::qreg_.initialize_omp(BaseState::threads_, omp_threshold_rank);
+}
+
+void State::initialize_qreg(uint_t num_qubits, const chstate_t &state)
+{
+  if(BaseState::qreg_.get_n_qubits() != num_qubits)
+  {
+    throw std::invalid_argument("CH::State::initialize: initial state does not match qubit number.");
+  }
+  BaseState::qreg_ = state;
+  BaseState::qreg_.initialize_omp(BaseState::threads_, omp_threshold_rank);
+}
+
+void State::set_config(const json_t &config)
+{
+  // Set the error upper bound in the stabilizer rank approximation
+  JSON::get_value(approximation_error, "ch_approximation_error", config);
+  // Set the number of samples used in the norm estimation routine
+  JSON::get_value(norm_estimation_samples, "ch_norm_estimation_samples", config);
+  // Set the number of steps used in the metropolis sampler before we
+  // consider the distribution as approximating the output
+  JSON::get_value(metropolis_mixing_steps, "ch_mixing_time", config);
+  //Set the threshold of the decomposition before we use omp
+  JSON::get_value(omp_threshold_rank, "ch_parallel_threshold", config);
+  //Set the truncation threshold for the probabilities snapshot.
+  JSON::get_value(snapshot_chop_threshold, "chop_threshold", config);
+  //Set the number of samples for the probabilities snapshot
+  JSON::get_value(probabilities_snapshot_samples, "probabilities_snapshot_samples", config);
+}
+
+std::pair<uint_t, uint_t> State::decomposition_parameters(const std::vector<Operations::Op> &ops)
+{
+  double xi=1.;
+  uint_t three_qubit_gate_count = 0;
+  for (const auto op: ops)
+  {
+    if (op.type == Operations::OpType::gate)
+    {
+      compute_extent(op, xi);
+      auto it = CHSimulator::gate_types_.find(op.name);
+      if (it->second == CHSimulator::Gatetypes::non_clifford && op.qubits.size() == 3)
+      { //We count the number of 3 qubit gates for normalisation purposes
+        three_qubit_gate_count++;
+      }
+    }
+  }
+  uint_t chi=1;
+  if (xi >1)
+  {
+    chi = std::ceil(xi/std::pow(approximation_error, 2));
+  }
+  return std::pair<uint_t, uint_t>({chi, three_qubit_gate_count});
+}
+
+std::pair<bool, size_t> State::check_stabilizer_opt(const std::vector<Operations::Op> &ops)
+{
+  for(auto op = ops.cbegin(); op != ops.cend(); op++)
```

Comment:
This is a good candidate for the new `for( : )` you used before:
```c++
for(auto& op : ops){
...
}
```

Reply:
I left this as is because make use of pointer arithmetic later in the function (returning `op - op.cbegin()`)